### PR TITLE
Ji/create clickhouse datasource

### PR DESCRIPTION
**Pull request by tzjames**
### Features and Changes

Creates an inbuilt Growthbook Clickhouse option

### Testing

- Create new datasource
- Click on the create in built datasource button
- Click Finish
- Edit the anonymous_id assignment query on the datasource page 
- Click Test Query
- See 4 results (`test_enriched_events` needs to have events that match the org in your database)

### Screenshots

Screenshot 2024-10-17 at 6 50 22 PM
when the sample database is also showing:
Screenshot 2024-10-17 at 6 50 01 PM

### CODE REVIEW

File: packages/back-end/src/services/clickhouse.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  for (let i = 3; i < length; i++) {
+    password += allChars[Math.floor(Math.random() * allChars.length)];
+  }
+
+  return password
+    .split("")
+    .sort(() => 0.5 - Math.random())
+    .join("");
+}
```
Comment (toots) [inline packages/back-end/src/services/clickhouse.ts:33]: We should consider using a third party here like this: https://www.npmjs.com/package/generate-password. This package uses the node crypto module to generate random bytes.

File: packages/front-end/components/Settings/NewDataSourceForm.tsx [hunk 2]
```diff
@@ -410,6 +410,26 @@ const NewDataSourceForm: FC<{
...
+      newDatasourceForm: true,
+    });
+
+    setCreatedDatasource(res.datasource);
+    createResources(res.datasource);
+    setStep("done");
+  };
+
+  // TODO: Make this a feature flag
+  const showInbuiltDatasource = true;
```
Comment (toots) [inline packages/front-end/components/Settings/NewDataSourceForm.tsx:431]: This should be made conditional.

**Review by toots — COMMENTED** [review]
This looks straight forward to me.

File: packages/back-end/src/services/clickhouse.ts (hunk 1 above)
Comment (romain-growthbook) [reply packages/back-end/src/services/clickhouse.ts:33]: We should consider using a third party here like this: https://www.npmjs.com/package/generate-password. This package uses the node crypto module to generate random bytes.

File: packages/front-end/components/Settings/NewDataSourceForm.tsx (hunk 2 above)
Comment (romain-growthbook) [reply packages/front-end/components/Settings/NewDataSourceForm.tsx:431]: This should be made conditional.

**Review by romain-growthbook — COMMENTED** [review]
This looks straight forward to me.

File: packages/back-end/src/app.ts [hunk 3]
```diff
@@ -657,6 +657,13 @@ app.post(
   datasourcesController.fetchBigQueryDatasets
 );
 
+if (IS_CLOUD) {
```
Comment (romain-growthbook) [inline packages/back-end/src/app.ts:660]: I think long term we should find a way to be more dev-friendly with those.

File: packages/back-end/types/datasource.d.ts [hunk 4]
```diff
@@ -12,6 +12,7 @@ import { MetricType } from "./metric";
 import { MssqlConnectionParams } from "./integrations/mssql";
 
 export type DataSourceType =
+  | "growthbook_clickhouse"
```
Comment (romain-growthbook) [inline packages/back-end/types/datasource.d.ts:15]: Aren't we here just overlooking the fact that this should be database type clickhouse, schema growthbook? Not blocking for now.

File: packages/front-end/pages/datasources/[did].tsx [hunk 5]
```diff
@@ -61,7 +61,10 @@ const DataSourcePage: FC = () => {
     (d && permissionsUtil.canDeleteDataSource(d) && !hasFileConfig()) || false;
 
   const canUpdateConnectionParams =
-    (d && permissionsUtil.canUpdateDataSourceParams(d) && !hasFileConfig()) ||
+    (d &&
+      permissionsUtil.canUpdateDataSourceParams(d) &&
+      !hasFileConfig() &&
+      d.type != "growthbook_clickhouse") ||
```
Comment (romain-growthbook) [inline packages/front-end/pages/datasources/[did].tsx:67]: These should be extracted and encapsulated.

**Review by romain-growthbook — APPROVED** [review]
Nothing blocking for a merge into the feature branch. Couple of things we might want to clean up before eventually merging to `main`.

**Pull request merged**